### ISSUE TITLE
Increase default GC window?

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/nd4j#2856]
Now that workspaces are enabled by default in DL4J, perhaps we should set default auto GC window to a larger value?
For example, wme are recommending 5 sec here: https://deeplearning4j.org/workspaces#garbage-collector

**Comment by raver119** [comment]
we should be cautious here. maybe set this value at mln/cg init method to
make sure nd4j isn't affected?

**Comment by AlexDBlack** [comment]
Yeah, I guess there's a lot of ways a default could go wrong. Though setting it in MLN/CG too might be risky in some cases. For example:

- User trains a net with workspaces enabled, then after trains one without -> maybe OOM?
- User has multiple threads... one with DL4J net (with WS) and another doing non-DL4J uses of ND4J -> OOM?

What about if we did things a bit more intelligently with the CG window? (Unless we do something like this already?)
Like we track the number (and/or number of bytes) of arrays allocated outside of workspaces since the last periodic GC. If that is below some threshold, we skip the periodic GC this iteration?
That way, a lower GC period used in conjunction with DL4J worspaces shouldn't be as much of an issue.

**Comment by raver119** [comment]
that might work actually. we know allocations. I'll take that.

**Comment by saudet** [comment]
JavaCPP is already tracking physical memory for CPU. We just need something for GPU really..
